lupa/_lupa.pyx

```diff
@@ -242,14 +253,22 @@ cdef class LuaRuntime:
     cdef object _attribute_getter
     cdef object _attribute_setter
     cdef bint _unpack_returned_tuples
+    cdef size_t _max_memory
+    cdef size_t _memory_left
 
     def __cinit__(self, encoding='UTF-8', source_encoding=None,
                   attribute_filter=None, attribute_handlers=None,
                   bint register_eval=True, bint unpack_returned_tuples=False,
-                  bint register_builtins=True, overflow_handler=None):
-        cdef lua_State* L = lua.luaL_newstate()
+                  bint register_builtins=True, overflow_handler=None,
+                  max_memory=0):
+        cdef lua_State* L
+
+        self._memory_left = 0
+
+        L = lua.lua_newstate(<lua.lua_Alloc>&_lua_alloc_restricted, &self._memory_left)
```

Comment:
> Do I understand correctly that the allocator function is not needed when no memory limit is used? Then we could keep the old call in the normal (no limit) case.
> I'd really like to get rid of the set_max_memory() method, if possible. It feels better to make this an init time only setting, because the effect of changing the maximum is really indirect. As your test shows, it only applies when the next allocation happens (which might be at unexpected times) and does not fail immediately, even though we might already have allocated more memory than the new limit.

I have to use the allocator function because `set_max_memory()` might be called in the future.

In my use case, I have to load multiple libraries just like the stdlib before running the actual code which should not count towards the memory limit of the code.

Maybe, the default could be None in which case the default `luaL_newstate` allocator will be used and `set_max_memory()` will throw? ([like how it was previously](https://github.com/scoder/lupa/pull/212/commits/2669b6346b574d502cfbbcf2b0f79d3d3bf32d0b#diff-cf9743b035eef33132b4cdec8aad44f5b0c2dfa113c858624a0bd3853e2b6604R254))

> BTW, why is it runtime.max_memory for reading but runtime.set_max_memory() for writing? Why not use a property for both?

Personal preference, and a relic from earlier testing stuff. (max_memory allocator was optional earlier and this function [threw an exception](https://github.com/scoder/lupa/pull/212/commits/bbfaf16b14d0d25a7dcb28d9eb8c50b81fdd3756#diff-cf9743b035eef33132b4cdec8aad44f5b0c2dfa113c858624a0bd3853e2b6604R486) and setting properties really shouldn't throw an exception imo)